### PR TITLE
red-knot: Add a method to resolve a file for an arbitrary `VfsPath`

### PR DESCRIPTION
**Pull request by MichaReiser**
## Summary

When I started working on porting the module resolver, I realized that the `path_to_module` function operates on an arbitrary `VfsPath` and it must be able to resolve it to a `VfsFile`. 

The problem I ran into is that the existing salsa db doesn't provide an API to resolve the file for an arbitrary path. This PR addresses this by replacing `db.file` and `db.vendored` with `vendored_path_to_file`, `file_system_path_to_file`, `vfs_path_to_file` functions.

The advantage of free-standing functions over trait functions is that these can take an `impl AsRef<.Path>` as argument which makes them more ergonomic. I kept the path specific operations because they're slightly more ergonomic (they don't require creating a `VfsPath` first).

## Test Plan

`cargo test`

### CODE REVIEW

File: crates/ruff_db/src/vfs.rs [hunk 1]
```diff
@@ -6,10 +6,47 @@ use dashmap::mapref::entry::Entry;
...
+use crate::vfs::private::FileStatus;
 use crate::{Db, FxDashMap};
 
 mod path;
 
+/// Interns a file system path and returns a salsa `File` ingredient.
+///
+/// Returns `None` if the path doesn't exist, isn't accessible, or if the path points to a directory.
+#[inline]
+pub fn file_system_path_to_file(db: &dyn Db, path: impl AsRef<FileSystemPath>) -> Option<VfsFile> {
```
Comment (AlexWaygood) [inline crates/ruff_db/src/vfs.rs:18]: I would consider naming this function `system_path_to_file`. It's more concise, and IMO having `file` twice in the name is a little confusing.

File: crates/ruff_db/src/vfs.rs [hunk 2]
```diff
@@ -6,10 +6,47 @@ use dashmap::mapref::entry::Entry;
...
+    // It's important that the `.file_system` method always returns a file and that the status is queried here
+    // to ensure that salsa knows which queries depend on the existence of this file, even if it doesn't exist
+    // yet or doesn't exist anymore.
+    match file.status(db) {
+        FileStatus::Exists => Some(file),
+        FileStatus::Deleted => None,
+    }
+}
+
+/// Interns a vendored file path. Returns `None` if no such vendored file exists and `Some` otherwise.
```
Comment (AlexWaygood) [inline crates/ruff_db/src/vfs.rs:31]:
```suggestion
/// Interns a vendored file path. Returns `Some(file)` if the vendored file exists, else `None`.
```

File: crates/ruff_db/src/vfs.rs [hunk 3]
```diff
@@ -6,10 +6,47 @@ use dashmap::mapref::entry::Entry;
...
+///
+/// Returns `None` if the path doesn't exist, isn't accessible, or if the path points to a directory.
+#[inline]
+pub fn file_system_path_to_file(db: &dyn Db, path: impl AsRef<FileSystemPath>) -> Option<VfsFile> {
+    let file = db.vfs().file_system(db, path.as_ref());
+
+    // Query the status of the file to determine if it exists.
+    // It's important that the `.file_system` method always returns a file and that the status is queried here
+    // to ensure that salsa knows which queries depend on the existence of this file, even if it doesn't exist
+    // yet or doesn't exist anymore.
```
[the remark this answers is not in the document]
Comment (MichaReiser) [reply crates/ruff_db/src/vfs.rs:24]: I rewrote the sentence

[the remark this answers is not in the document]
Comment (AlexWaygood) [reply crates/ruff_db/src/vfs.rs]: Thank you! That's much clearer now

File: crates/ruff_db/src/vfs.rs [hunk 4]
```diff
@@ -6,10 +6,47 @@ use dashmap::mapref::entry::Entry;
...
+/// Interns a file system path and returns a salsa `File` ingredient.
+///
+/// Returns `None` if the path doesn't exist, isn't accessible, or if the path points to a directory.
+#[inline]
+pub fn system_path_to_file(db: &dyn Db, path: impl AsRef<FileSystemPath>) -> Option<VfsFile> {
+    let file = db.vfs().file_system(db, path.as_ref());
+
+    // It's important that `vfs.file_system` creates a `VfsFile` even for files that don't exist or don't
+    // exist anymore so that Salsa can track that the caller of this function depends on the existence of
+    // that file. This function filters out file that don't exist, but Salsa will know that it must
```
Comment (AlexWaygood) [inline crates/ruff_db/src/vfs.rs:23]:
```suggestion
    // that file. This function filters out files that don't exist, but Salsa will know that it must
```

**Review by AlexWaygood — APPROVED** [review]

**Pull request merged**